Provide IlValue as condition for the Switch IlBuilder service

nbhuiyan opened the pull request:
The existing Switch and  TableSwitch (and their overloads) services
could only be used if the condition to switch upon was a named
variable. It was not possible to provide an IlValue to switch on,
which limited the scenarios where we could use the Switch service.

This changeset adds 2 overloads of the Switch service to use IlValue
as the condition to switch on, while still ensuring that the IlValue
provided is of Int32 data type.

Signed-off-by: Nazim Uddin Bhuiyan <nazimudd@ualberta.ca>

This changeset was motivated by the [lljb project](https://github.com/nbhuiyan/lljb).

mstoodle commented on compiler/ilgen/OMRIlBuilder.cpp:
```diff
@@ -2604,7 +2604,29 @@ OMR::IlBuilder::Switch(const char *selectionVar,
                   uint32_t numCases,
                   JBCase **cases)
    {
-   TR::IlValue *selectorValue = Load(selectionVar);
```
To be honest, I can no longer remember why we implemented `Switch` taking a condition "variable" and not just an `IlValue` :( . But I'm sure it's being used, so you're right that we shouldn't just remove it.

nbhuiyan commented:
@mstoodle In addressing your review comments, I felt that `api-gen` is not going to be able to correctly handle these additional overloads of the switch services in languages that do not support overloads. Here are the two entries for for Switch service:
```json
               { "name": "Switch"
                , "overloadsuffix": "WithArgArray"
                , "flags": []
                , "return": "none"
                , "parms": [
                    {"name":"selectionVar","type":"constString"},
                    {"name":"defaultBuilder","type":"IlBuilder","attributes":["in_out"]},
                    {"name":"numCases","type":"int32"},
                    {"name":"cases","type":"JBCase","attributes":["array","can_be_vararg"],"array-len":"numCases"}
                    ]
                },
                { "name": "Switch"
                    , "overloadsuffix": "ValWithArgArray"
                    , "flags": []
                    , "return": "none"
                    , "parms": [
                        {"name":"selectorValue","type":"IlValue"},
                        {"name":"defaultBuilder","type":"IlBuilder","attributes":["in_out"]},
                        {"name":"numCases","type":"int32"},
                        {"name":"cases","type":"JBCase","attributes":["array","can_be_vararg"],"array-len":"numCases"}
                        ]
                },

```

I believe that the four different names that will be generated for a language without overload support are
* `Switch`
* `SwitchWithArgArray`
* `Switch`
* `SwitchValWithArgArray`

I have already discussed this problem with @Leonardo2718, and would like your thoughts on this.

mstoodle commented:
Yeah, it's awkward, for sure. I would actually prefer that `Switch` take a value and the other one (taking a variable name) be phased out because I think it's more natural for these services to consume an `IlValue`. I cannot actually remember a reason why they started out taking a variable name :( .

@charliegracie @rwy0717 @Leonardo2718 @aviansie-ben @jduimovich I think you're the most likely people to already be using `Switch` or `TableSwitch`. Would you be opposed if we changed the two `Switch` services so that the input selector is an `IlValue` rather than the current `const char *` variable name ?

We welcome feedback from anyone who has an opinion on this topic, of course, so please don't hold back your comments if I didn't explicitly @ mention you :) .

Leonardo2718 commented:
I'm fine with changing the API to take an IlValue instead of a variable name.

nbhuiyan commented:
I think enough time has elapsed without any opposition to removing the existing versions of the Switch/TableSwitch services for me to move forward with this change.

mstoodle commented:
@genie-omr build all

0xdaryl commented:
@genie-omr build zos

0xdaryl commented:
The zOS failure (from December!) was infrastructural.  All tests pass.  @mstoodle, do you approve merging?

mstoodle commented on jitbuilder/apigen/jitbuilder.api.json:
```diff
@@ -1464,27 +1464,27 @@
-                    {"name":"numCases","type":"int32"},
-                    {"name":"cases","type":"JBCase","attributes":["array","can_be_vararg"],"array-len":"numCases"}
-                    ]
+                    , "overloadsuffix": "WithArgArray"
```
also extra indent here

mstoodle commented on jitbuilder/apigen/jitbuilder.api.json:
```diff
@@ -1464,27 +1464,27 @@
-                    {"name":"numCases","type":"int32"},
-                    {"name":"cases","type":"JBCase","attributes":["array","can_be_vararg"],"array-len":"numCases"}
-                    ]
+                    , "overloadsuffix": "WithArgArray"
```
I'll merge this now, but please fix the indentation as a follow-on PR :) .
